File: mazes-for-programmers/python/src/mazes/wilsons.py
from random import choice

from image_saver import save
from colored_grid import ColoredGrid


class Wilsons:
    @staticmethod
    def on(grid):
        unvisited = list(grid.each_cell())
        first = choice(unvisited)
        unvisited.remove(first)
        while unvisited:
            cell = choice(unvisited)
            path = [cell]
            while cell in unvisited:
                cell = choice(cell.neighbors())
                position = path.index(cell) if cell in path else None
                if position is not None:
                    path = path[0 : position + 1]
                else:
                    path.append(cell)
            for index in range(len(path) - 1):
                path[index].link(path[index + 1])
                unvisited.remove(path[index])
        return grid


def wilsons_demo():
    grid = ColoredGrid(11, 11)
    Wilsons.on(grid)
    save(grid.to_img(), "wilsons.png")

    middle = grid[grid.rows // 2, grid.columns // 2]
    grid.set_distances(middle.distances())

    save(grid.to_img(), "wilsons_colored.png")


if __name__ == "__main__":
    wilsons_demo()

File: mazes-for-programmers/kotlin/src/main/kotlin/io/github/ocirne/mazes/algorithms/Wilsons.kt
package io.github.ocirne.mazes.algorithms

import io.github.ocirne.mazes.grids.Cell
import io.github.ocirne.mazes.grids.GridProvider
import io.github.ocirne.mazes.grids.MutableGrid

class Wilsons : PassageCarver {

    override fun on(gridProvider: GridProvider, startAt: (MutableGrid) -> Cell): MutableGrid {
        val grid = gridProvider.createPassageCarver()
        val unvisited = grid.eachCell().toMutableList()
        val first = unvisited.random()
        unvisited.remove(first)
        while (unvisited.isNotEmpty()) {
            var cell = unvisited.random()
            var path = mutableListOf(cell)
            while (unvisited.contains(cell)) {
                cell = cell.neighbors().random()
                val position = path.indexOf(cell)
                if (position >= 0)
                    path = path.subList(0, position + 1)
                else
                    path.add(cell)
            }
            path.zipWithNext { v, w -> v.link(w) }
            unvisited.removeAll(path.subList(0, path.size-1).toSet())
        }
        return grid
    }
}
